spark-3.1-spanner-lib/src/test/java/com/google/cloud/spark/spanner/ReadIntegrationTestBase.java

```diff
@@ -58,39 +60,191 @@ public void testReadFromCompositeTable() {
+    //        are both equal to:
+    //   2023-01-01
+    //
+    //   2. milliseconds: 1693551600000 and 1693612800000
+    //        are both equal to:
+    //   2023-09-01
+    List<String> gotHSStr = new ArrayList();
+    gotHs.forEach(ts -> gotHSStr.add(ts.toString()));
+    List<String> wantHSStr = new ArrayList();
+    wantHs.forEach(ts -> wantHSStr.add(ts.toString()));
+    assertThat(gotHSStr).containsExactlyElementsIn(wantHSStr);
+
+    // 10. Validate I field ARRAY<TIMESTAMP> values.
+    List<Row> allIs = df.select("I").collectAsList();
+    assertEquals(allIs.size(), 2);
+    Row row1 = allIs.get(0);
+    assertEquals(row1.size(), 1);
+    Row row2 = allIs.get(1);
+    assertEquals(row2.size(), 1);
+    List<Timestamp> gotIs = new ArrayList();
+    gotIs.addAll(row1.getList(0));
+    gotIs.addAll(row2.getList(0));
+    Collections.sort(gotIs);
+
+    List<Timestamp> wantIs = new ArrayList();
+    Arrays.asList(
+            ZonedDateTime.parse("2023-08-26T12:11:10Z"),
+            ZonedDateTime.parse("2023-08-27T12:11:09Z"),
+            ZonedDateTime.parse("2023-09-22T12:11:10Z"),
+            ZonedDateTime.parse("2023-09-23T12:11:09Z"))
+        .forEach(zd -> wantIs.add(new Timestamp(zd.toInstant().toEpochMilli())));
+    Collections.sort(wantIs);
+    assertThat(gotIs).containsExactlyElementsIn(wantIs);
+  }
+
+  @Test
+  public void testNumericLimits() {
+    Dataset<Row> df = readFromTable("valueLimitsTable");
+    assertThat(df.count()).isEqualTo(4);
+
+    // 1. Test the limits of INT64.
+    List<Long> gotAs = df.select("A").as(Encoders.LONG()).collectAsList();
+    Collections.sort(gotAs);
+
+    long[] lA = {Long.MIN_VALUE, Long.MAX_VALUE, 0, 1};
+    List<Long> wantAs = Arrays.stream(lA).boxed().collect(Collectors.toList());
+    Collections.sort(wantAs);
+    assertThat(wantAs).containsExactlyElementsIn(gotAs);
+
+    // 2. Test the limits of FLOAT64.
+    List<Double> gotBs = df.select("B").as(Encoders.DOUBLE()).collectAsList();
+    Collections.sort(gotBs);
+
+    List<Double> wantBs =
+        Arrays.asList(Double.NaN, Double.POSITIVE_INFINITY, Double.NEGATIVE_INFINITY, 0.657818);
+    Collections.sort(wantBs);
+    assertThat(wantBs).containsExactlyElementsIn(gotBs);
+
+    // 3. Test the limits of NUMERIC.
+    List<BigDecimal> gotCs = df.select("C").as(Encoders.DECIMAL()).collectAsList();
+    Collections.sort(gotCs);
+
+    List<BigDecimal> wantCs =
+        Arrays.asList(
+            asSparkBigDecimal("-9.9999999999999999999999999999999999999E+28"),
+            asSparkBigDecimal("9.9999999999999999999999999999999999999E+28"),
+            asSparkBigDecimal("10.389"),
+            asSparkBigDecimal("-10.389"));
+    Collections.sort(wantCs);
+    assertThat(wantCs).containsExactlyElementsIn(gotCs);
+  }
+
+  @Test
+  public void testReadJSON() {
+    Dataset<Row> df = readFromTable("players");
+    long totalRows = df.count();
+    assertThat(totalRows).isEqualTo(2);
+    if (true) {
```

Comment:
If the test is not passing, we should have a fix in the pull request. Otherwise, we should remove the testReadJSON. It doesn't make sense to have the if(true) return logic here.

Reply:
I put a TODO down below it explaining why, but sure I can remove it entirely.

Reply:
I thought we are going to support JSON in the initial launch. What's the error we got ?

Reply:
We def do support it and I've verified it in actual example.py, the test is just a silly one in our test we construct a struct but when comparing the output the format is different and I wasted so much time on it it isn't worth it fumbling with Java's comparisons meanwhile there are much higher hanging fruit, hence why I had put that TODO:

```shell
Want: ["{\"a\":"b", 1:2}", "{\"1\":1,\"2\":2}"]
Got:    [{"a":"b", 1:2}, {"1":1,"2":2}]
```